File: ports/rust/src/core/global/bcsstring.rs
// BcsString represents the core string type, mapped to Rust's native String
pub struct BcsStringWrapper(String);

impl BcsStringWrapper {
    pub fn new(val: &str) -> Self {
        BcsStringWrapper(val.to_string())
    }

    pub fn to_upper(&self) -> String {
        self.0.to_uppercase()
    }

    pub fn to_lower(&self) -> String {
        self.0.to_lowercase()
    }
}

impl std::fmt::Display for BcsStringWrapper {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}

impl BcsStringWrapper {
    pub fn length(&self) -> usize {
        self.0.len()
    }

    pub fn is_empty(&self) -> bool {
        self.0.is_empty()
    }

    pub fn substr(&self, index: usize, length: usize) -> String {
        let str = &self.0;
        if index >= str.len() {
            return String::new();
        }
        let end = std::cmp::min(index + length, str.len());
        str[index..end].to_string()
    }

    pub fn to_utf8(&self) -> Vec<u8> {
        self.0.as_bytes().to_vec()
    }
}

File: ports/go/src/core/global/bcsstring.go
package global

import "strings"

// BcsString represents the core string type, mapped to Go's native string
type BcsStringUtf8 string

// NewBcsString creates a new BcsString
func NewBcsString(val string) BcsString {
	return BcsString(val)
}

// ToUpper converts the string to uppercase
func (s BcsString) ToUpper() BcsString {
	return BcsString(strings.ToUpper(string(s)))
}

// ToLower converts the string to lowercase
func (s BcsString) ToLower() BcsString {
	return BcsString(strings.ToLower(string(s)))
}

// Length returns the number of characters in the string
func (s BcsString) Length() int {
	return len(s)
}

// IsEmpty checks if the string has a length of 0
func (s BcsString) IsEmpty() bool {
	return len(s) == 0
}

// Substr returns a substring starting at index with a given length
func (s BcsString) Substr(index, length int) BcsString {
    str := string(s)
    if index < 0 || index >= len(str) {
        return ""
    }
    end := index + length
    if end > len(str) {
        end = len(str)
    }
	return BcsString(str[index:end])
}

// ToUtf8 converts the BcsString to a standard UTF-8 byte slice
func (s BcsString) ToUtf8() []byte {
	return []byte(s)
}
